controllers/tls_workflow.go

```diff
@@ -133,24 +126,124 @@ func LinkGatewayToClusterIssuerFunc(objs controller.Store) machinery.LinkFunc {
 				return p.Spec.IssuerRef.Name == clusterIssuer.GetName() && p.Spec.IssuerRef.Kind == certmanagerv1.ClusterIssuerKind
 			})
 
-			if len(linkedPolicies) == 0 {
-				return nil
+			return findLinkedGatewaysForIssuer(linkedPolicies, gateways)
+		},
+	}
+}
+
+func findLinkedGatewaysForIssuer(linkedPolicies []*kuadrantv1alpha1.TLSPolicy, gateways []*gwapiv1.Gateway) []machinery.Object {
+	if len(linkedPolicies) == 0 {
+		return nil
+	}
+
+	// Can infer linked gateways through the policy
+	linkedGateways := lo.Filter(gateways, func(g *gwapiv1.Gateway, index int) bool {
+		for _, l := range linkedPolicies {
+			if string(l.Spec.TargetRef.Name) == g.GetName() && g.GetNamespace() == l.GetNamespace() {
+				return true
+			}
+		}
+
+		return false
+	})
+
+	return lo.Map(linkedGateways, func(item *gwapiv1.Gateway, index int) machinery.Object {
+		return &machinery.Gateway{Gateway: item}
+	})
+}
+
+// Common functions used across multiple reconcilers
+
+func IsTLSPolicyValid(ctx context.Context, s *sync.Map, policy *kuadrantv1alpha1.TLSPolicy) (bool, error) {
+	logger := controller.LoggerFromContext(ctx).WithName("IsPolicyValid")
+
+	store, ok := s.Load(TLSPolicyAcceptedKey)
+	if !ok {
+		logger.V(1).Info("TLSPolicyAcceptedKey not found, policies will be checked for validity by current status")
+		return meta.IsStatusConditionTrue(policy.Status.Conditions, string(gatewayapiv1alpha2.PolicyReasonAccepted)), nil
+	}
+
+	isPolicyValidErrorMap := store.(map[string]error)
+
+	return isPolicyValidErrorMap[policy.GetLocator()] == nil, isPolicyValidErrorMap[policy.GetLocator()]
+}
+
+func GetTLSPoliciesByEvents(topology *machinery.Topology, events []controller.ResourceEvent) []machinery.Policy {
+	policies := lo.Filter(topology.Policies().Items(), func(item machinery.Policy, index int) bool {
+		_, ok := item.(*kuadrantv1alpha1.TLSPolicy)
+		return ok
+	})
+
+	var affectedPolicies []machinery.Policy
+	for _, event := range events {
+		if event.Kind == machinery.GatewayGroupKind {
+			ob := event.NewObject
+			if ob == nil {
+				ob = event.OldObject
 			}
 
-			// Can infer linked gateways through the policy
-			linkedGateways := lo.Filter(gateways, func(g *gwapiv1.Gateway, index int) bool {
-				for _, l := range linkedPolicies {
-					if string(l.Spec.TargetRef.Name) == g.GetName() && g.GetNamespace() == l.GetNamespace() {
+			g := machinery.Gateway{Gateway: ob.(*gwapiv1.Gateway)}
+
+			affectedPolicies = append(affectedPolicies, lo.Filter(policies, func(item machinery.Policy, index int) bool {
+				for _, tg := range item.GetTargetRefs() {
+					if g.GetLocator() == tg.GetLocator() {
```

Comment:
Have you considered using [`g.Policies()`](https://github.com/Kuadrant/policy-machinery/blob/72b15c675f8d03489ae926288b3fd2d426174fe6/machinery/types.go#L39C2-L39C12)?

Reply:
Hmm, dont think i'm able to since `g` is a new `machinery.Gateway` instance, so the `policies` will be `nil` unless you're saying to get the `machinery.Gateway` instance elsewhere? 🤔

Reply:
True. I was thinking on getting the gateway from the topology. Indeed `g := machinery.Gateway{Gateway: ob.(*gwapiv1.Gateway)}` is no good.

Reply:
Reverted this filtering